src/macaron/dependency_analyzer/java_repo_finder.py

```diff
@@ -0,0 +1,194 @@
+    return urls
+
+
+def parse_gav(gav: str) -> tuple[str, str, str]:
+    """
+    Parse the passed GAV string, splitting it into its constituent parts.
+
+    Parameters
+    ----------
+    gav : str
+        An artifact represented as a GAV (group, artifact, version) in the format: G:A:V.
+
+    Returns
+    -------
+    tuple[str, str, str]
+        The tuple of separated GAV components.
+    """
+    split = gav.split(":")
+    if len(split) < 3:
+        logger.debug("Could not parse GAV: %s", gav)
+        return "", "", ""
+
+    group, artifact, version = split
+    group = group.replace(".", "/")
+    return group, artifact, version
+
+
+def retrieve_pom(session: requests.Session, url: str) -> str:
+    """
+    Attempt to retrieve the file located at the passed URL using the passed Session.
+
+    Parameters
+    ----------
+    session : requests.Session
+        The HTTP session to use for attempting the GET request.
+    url : str
+        The URL for the GET request.
+
+    Returns
+    -------
+    str :
+        The retrieved file data or an empty string.
+    """
+    if not url.endswith(".pom"):
+        return ""
+
+    try:
+        res = session.get(url)
+    except (requests.RequestException, OSError) as error:
+        logger.debug("Error during pom retrieval: %s", error)
+        return ""
+
+    if not res.ok:
+        logger.warning("Failed to retrieve pom from: %s, error code: %s", url, res.status_code)
+        return ""
+
+    logger.debug("Found artifact POM at: %s", url)
+
+    return res.text
+
+
+def _find_element(parent: typing.Optional[Element], target: str) -> typing.Optional[Element]:
+    if parent is None:
+        return None
+
+    # Attempt to match the target tag within the children of parent.
+    for child in parent:
+        # Account for raw tags, and tags accompanied by Maven metadata enclosed in curly braces. E.g. '{metadata}tag'
+        if child.tag == target or child.tag.endswith("}" + target):
+            return child
+    return None
+
+
+def parse_pom(pom: str, tags: list[str]) -> Iterator[str]:
+    """
+    Parse the passed pom and extract the passed tags.
+
+    Parameters
+    ----------
+    pom : str
+        The POM as a string
+    tags : list[str]
+        The list of tags to try extracting from the POM
+
+    Returns
+    -------
+    Iterator[str] :
+        The extracted contents of any matches tags
+    """
+    try:
+        xml: Element = fromstring(pom)
+    except defusedxml.ElementTree.ParseError as error:
+        logger.debug("Failed to parse XML: %s", error)
+        return iter([])
+
+    results = []
+
+    # Try to match each tag with the contents of the POM.
+    for tag in tags:
+        element: typing.Optional[Element] = xml
+        tag_parts = tag.split(".")
+        for index, tag_part in enumerate(tag_parts):
+            element = _find_element(element, tag_part)
+            if element is None:
```

Comment:
```suggestion
            if not element:
```

Reply:
Interestingly, it turns out these are not equivalent in this case, as using `if not element` leads to different results.

Reply:
We shouldn't replace ` if element is None` with `if not element` if we need to check that `element` is `None`. Here if `element` is an empty list, `if not element` would still be `False`, even though it's not `None. See https://docs.python.org/3/library/stdtypes.html#truth-value-testing.

Reply:
Hmmm. I see that `_find_element` returned `Optional[Element]` so checking it with `if not element` would be okay. However, because `Element` implement its `__bool__` :
```python
    def __bool__(self):
        warnings.warn(
            "The behavior of this method will change in future versions.  "
            "Use specific 'len(elem)' or 'elem is not None' test instead.",
            FutureWarning, stacklevel=2
            )
        return len(self._children) != 0 # emulate old behaviour, for now
```
This won't return the behavior that I was expecting (if the method return an `Element` instance but without any children `not element -> True` . I think we shouldn't change it to `if not element`).